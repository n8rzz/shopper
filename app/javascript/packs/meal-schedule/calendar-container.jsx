import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { CalendarDay } from './calendar-day.component';

export class CalendarContainer extends React.Component {
    constructor(props) {
        super(props);

        this._weekdayShortNames = moment.weekdaysShort();
    }

    _getfirstDayOfCurrentMonth() {
        return this.props.currentDate
            .startOf('month')
            .format('d');
    }

    _getCurrentMonthName() {
        return moment(this.props.currentDate).format('MMM');
    }

    _getCurrentYear() {
        return moment(this.props.currentDate).format('YYYY');
    }

    _findMealScheduleForDay(day = []) {
        if (day.length === 0) {
            return [];
        }

        return this.props.mealSchedules.filter(
            (schedule) => day.format('DD/MM') === moment(schedule.schedule_date).format('DD/MM'),
        );
    }

    _hasEventOnDay(day) {
        const foundEvent = this._findMealScheduleForDay(day);

        return typeof foundEvent.length > 0;
    }

    _buildTableHeadJsx() {
        return this._weekdayShortNames.map((day) => (
            <th
                className={'calendarTable-th'}
                key={`weekdayHeader-${day}`}
            >
                {day}
            </th>
        ));
    }

    _buildBlankCellJsx(numberOfCells = -1) {
        const blankCellListJsx = [];
        const cellCount = numberOfCells !== -1
            ? numberOfCells
            : this._getfirstDayOfCurrentMonth();

        // eslint-disable-next-line no-plusplus
        for (let i = 0; i < cellCount; i++) {
            const blankCellJsx = (
                <td
                    key={`blankDay-${i}`}
                    className={'calendarTable-td'}
                >
                    <div className={'calendarTable-td-cell mix-calendarTable-td-cell_isEmpty'}>
                        &nbsp;
                    </div>
                </td>
            );

            blankCellListJsx.push(blankCellJsx);
        }

        return blankCellListJsx;
    }

    _buildCalendarCellsJsx() {
        const calendarDaysJsx = [];

        // eslint-disable-next-line no-plusplus
        for (let dayNumber = 1; dayNumber <= this.props.currentDate.daysInMonth(); dayNumber++) {
            const rawDate = new Date(this.props.currentDate.year(), this.props.currentDate.month(), dayNumber);
            const eventDate = moment(rawDate);
            // FIXME: could be multiple, fix this
            const mealScheduleForDay = this._findMealScheduleForDay(eventDate);

            const calendarDayJsx = (
                <CalendarDay
                    key={`calendarDay-${dayNumber}`}
                    dayNumber={dayNumber}
                    mealScheduleForDay={mealScheduleForDay}
                    onClickDayHandler={this.props.onClickDayHandler}
                />
            );

            calendarDaysJsx.push(calendarDayJsx);
        }

        return calendarDaysJsx;
    }

    _buildTableBodyJsx() {
        const blankDaysJsx = this._buildBlankCellJsx();
        const daysInMonth = this._buildCalendarCellsJsx();
        const totalSlots = [...blankDaysJsx, ...daysInMonth];
        const rows = [];
        let cells = [];

        // FIXME: rework this, there should be a better way
        totalSlots.forEach((row, i) => {
            if (i % 7 === 0) {
                // when reach next week we contain all td in last week to rows
                rows.push(cells);
                cells = [];
                // in current loop we still push current row to new container
                cells.push(row);
            } else {
                cells.push(row);
            }

            if (i === totalSlots.length - 1) {
                const remainingEmptyCells = 7 - cells.length;

                rows.push([
                    ...cells,
                    ...this._buildBlankCellJsx(remainingEmptyCells),
                ]);
            }
        });

        return rows.map((rowJsx, i) => (<tr key={`calendar-row-${i}`}>{rowJsx}</tr>));
    }

    render() {
        return (
            <div className={'calendar'}>
                <div className={'calendar-hd'}>
                    <div className={'calendar-hd-btn'}>
                        <button
                            type={'button'}
                            className={'btn'}
                            onClick={this.props.onClickPreviousMonthHandler}
                        >
                            &lt;
                        </button>
                    </div>
                    <div className={'calendar-hd-title'}>
                        {this._getCurrentMonthName()} - {this._getCurrentYear()}
                    </div>
                    <div className={'calendar-hd-btn'}>
                        <button
                            type={'button'}
                            className={'btn'}
                            onClick={this.props.onClickNextMonthHandler}
                        >
                            &gt;
                        </button>
                    </div>
                </div>
                <div className={'calendar-bd'}>
                    <table className={'calendarTable'}>
                        <thead>
                            <tr>{this._buildTableHeadJsx()}</tr>
                        </thead>
                        <tbody>{this._buildTableBodyJsx()}</tbody>
                    </table>
                </div>
            </div>
        );
    }
}

CalendarContainer.propTypes = {
    currentDate: PropTypes.object.isRequired,
    mealSchedules: PropTypes.array.isRequired,
    onClickPreviousMonthHandler: PropTypes.func.isRequired,
    onClickNextMonthHandler: PropTypes.func.isRequired,
    onClickDayHandler: PropTypes.func.isRequired,
};
