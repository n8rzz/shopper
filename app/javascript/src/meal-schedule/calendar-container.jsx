import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import classnames from 'classnames';
import { CalendarDay } from './calendar-day.component';

export class CalendarContainer extends React.Component {
    constructor(props) {
        super(props);

        this._today = moment(new Date()).startOf('day');
        this._weekdayShortNames = moment.weekdaysShort();
        this._onClickCalendarHeaderHandler = this._onClickCalendarHeader.bind(this);
        this.state = {
            isCalendarBodyHidden: false,
        };
    }

    componentWillUnmount() {
        this._onClickCalendarHeaderHandler = null;
    }

    _getfirstDayOfCurrentMonth() {
        return this.props.currentDate
            .startOf('month')
            .format('d');
    }

    _getCurrentMonthName() {
        return moment.utc(this.props.currentDate).format('MMM');
    }

    _getCurrentYear() {
        return moment.utc(this.props.currentDate).format('YYYY');
    }

    _findMealScheduleForDay(day = []) {
        if (day.length === 0) {
            return [];
        }

        return this.props.mealSchedules.filter(
            (schedule) => day.format('DD/MM') === moment.utc(schedule.schedule_date).format('DD/MM'),
        );
    }

    _isEventDayToday(eventDate) {
        return eventDate.clone().startOf('day').isSame(this._today, 'd');
    }

    _hasEventOnDay(day) {
        const foundEvent = this._findMealScheduleForDay(day);

        return typeof foundEvent.length > 0;
    }

    _buildTableBodyClassnames() {
        return classnames({
            'calendar-bd': true,
            'u-isVisuallyHidden': this.state.isCalendarBodyHidden,
        });
    }

    _buildTableHeadJsx() {
        return this._weekdayShortNames.map((day) => (
            <th
                className={'calendarTable-th'}
                key={`weekdayHeader-${day}`}
            >
                {day}
            </th>
        ));
    }

    _buildBlankCellJsx(numberOfCells = -1) {
        const blankCellListJsx = [];
        const cellCount = numberOfCells !== -1
            ? numberOfCells
            : this._getfirstDayOfCurrentMonth();

        // eslint-disable-next-line no-plusplus
        for (let i = 0; i < cellCount; i++) {
            const blankCellJsx = (
                <td
                    key={`blankDay-${i}`}
                    className={'calendarTable-td'}
                >
                    <div className={'calendarTable-td-cell mix-calendarTable-td-cell_isEmpty'}>
                        &nbsp;
                    </div>
                </td>
            );

            blankCellListJsx.push(blankCellJsx);
        }

        return blankCellListJsx;
    }

    _buildCalendarCellsJsx() {
        const calendarDaysJsx = [];

        // eslint-disable-next-line no-plusplus
        for (let dayNumber = 1; dayNumber <= this.props.currentDate.daysInMonth(); dayNumber++) {
            const rawDate = new Date(this.props.currentDate.year(), this.props.currentDate.month(), dayNumber);
            const eventDate = moment.utc(rawDate);
            // FIXME: could be multiple, fix this
            const mealScheduleForDay = this._findMealScheduleForDay(eventDate);

            const calendarDayJsx = (
                <CalendarDay
                    key={`calendarDay-${dayNumber}`}
                    isToday={this._isEventDayToday(eventDate)}
                    dayNumber={dayNumber}
                    mealScheduleForDay={mealScheduleForDay}
                    onClickDayHandler={this.props.onClickDayHandler}
                />
            );

            calendarDaysJsx.push(calendarDayJsx);
        }

        return calendarDaysJsx;
    }

    _buildTableBodyJsx() {
        const blankDaysJsx = this._buildBlankCellJsx();
        const daysInMonth = this._buildCalendarCellsJsx();
        const totalSlots = [...blankDaysJsx, ...daysInMonth];
        const rows = [];
        let cells = [];

        // FIXME: rework this, there should be a better way
        totalSlots.forEach((row, i) => {
            if (i % 7 === 0) {
                // when reach next week we contain all td in last week to rows
                rows.push(cells);
                cells = [];
                // in current loop we still push current row to new container
                cells.push(row);
            } else {
                cells.push(row);
            }

            if (i === totalSlots.length - 1) {
                const remainingEmptyCells = 7 - cells.length;

                rows.push([
                    ...cells,
                    ...this._buildBlankCellJsx(remainingEmptyCells),
                ]);
            }
        });

        return rows.map((rowJsx, i) => (<tr key={`calendar-row-${i}`}>{rowJsx}</tr>));
    }

    _onClickCalendarHeader() {
        this.setState((prevState) => ({ isCalendarBodyHidden: !prevState.isCalendarBodyHidden }));
    }

    render() {
        return (
            <div className={'calendar'}>
                <div className={'calendar-hd'}>
                    <div className={'calendar-hd-btn'}>
                        <button
                            type={'button'}
                            className={'btn'}
                            onClick={this.props.onClickPreviousMonthHandler}
                        >
                            &lt;
                        </button>
                    </div>
                    <div
                        className={'calendar-hd-title'}
                        onClick={this._onClickCalendarHeaderHandler}
                    >
                        {this._getCurrentMonthName()} - {this._getCurrentYear()}
                    </div>
                    <div className={'calendar-hd-btn'}>
                        <button
                            type={'button'}
                            className={'btn'}
                            onClick={this.props.onClickNextMonthHandler}
                        >
                            &gt;
                        </button>
                    </div>
                </div>
                <div className={this._buildTableBodyClassnames()}>
                    <table className={'calendarTable'}>
                        <thead>
                            <tr>{this._buildTableHeadJsx()}</tr>
                        </thead>
                        <tbody>{this._buildTableBodyJsx()}</tbody>
                    </table>
                </div>
            </div>
        );
    }
}

CalendarContainer.propTypes = {
    currentDate: PropTypes.object.isRequired,
    mealSchedules: PropTypes.array.isRequired,
    onClickPreviousMonthHandler: PropTypes.func.isRequired,
    onClickNextMonthHandler: PropTypes.func.isRequired,
    onClickDayHandler: PropTypes.func.isRequired,
};
