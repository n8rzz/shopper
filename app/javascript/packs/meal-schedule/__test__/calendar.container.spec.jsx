import React from 'react';
import moment from 'moment';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import { CalendarContainer } from '../calendar-container';
import {
    currentDateMock,
    mealScheduleListMock,
} from '../../__mocks__/mocks';

const currentDateFixture = moment(currentDateMock);

describe('CalendarContainer', () => {
    test('renders correctly with valid props', () => {
        const component = shallow(<CalendarContainer
            currentDate={currentDateFixture}
            mealSchedules={mealScheduleListMock}
            onClickPreviousMonthHandler={sinon.spy()}
            onClickNextMonthHandler={sinon.spy()}
            onClickDayHandler={sinon.spy()}
        />);

        expect(component).toMatchSnapshot();
    });
});
