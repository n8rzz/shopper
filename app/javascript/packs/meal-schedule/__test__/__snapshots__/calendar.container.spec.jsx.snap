// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarContainer renders correctly with valid props 1`] = `
<div
  className="calendar"
>
  <div
    className="calendar-hd"
  >
    <div
      className="calendar-hd-btn"
    >
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        &lt;
      </button>
    </div>
    <div
      className="calendar-hd-title"
    >
      Sep
       - 
      2019
    </div>
    <div
      className="calendar-hd-btn"
    >
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        &gt;
      </button>
    </div>
  </div>
  <div
    className="calendar-bd"
  >
    <table
      className="calendarTable"
    >
      <thead>
        <tr>
          <th
            className="calendarTable-th"
            key="weekdayHeader-Sun"
          >
            Sun
          </th>
          <th
            className="calendarTable-th"
            key="weekdayHeader-Mon"
          >
            Mon
          </th>
          <th
            className="calendarTable-th"
            key="weekdayHeader-Tue"
          >
            Tue
          </th>
          <th
            className="calendarTable-th"
            key="weekdayHeader-Wed"
          >
            Wed
          </th>
          <th
            className="calendarTable-th"
            key="weekdayHeader-Thu"
          >
            Thu
          </th>
          <th
            className="calendarTable-th"
            key="weekdayHeader-Fri"
          >
            Fri
          </th>
          <th
            className="calendarTable-th"
            key="weekdayHeader-Sat"
          >
            Sat
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="calendar-row-0"
        />
        <tr
          key="calendar-row-1"
        >
          <CalendarDay
            dayNumber={1}
            key="calendarDay-1"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={2}
            key="calendarDay-2"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={3}
            key="calendarDay-3"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={4}
            key="calendarDay-4"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={5}
            key="calendarDay-5"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={6}
            key="calendarDay-6"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={7}
            key="calendarDay-7"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
        </tr>
        <tr
          key="calendar-row-2"
        >
          <CalendarDay
            dayNumber={8}
            key="calendarDay-8"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={9}
            key="calendarDay-9"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={10}
            key="calendarDay-10"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={11}
            key="calendarDay-11"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={12}
            key="calendarDay-12"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={13}
            key="calendarDay-13"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={14}
            key="calendarDay-14"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
        </tr>
        <tr
          key="calendar-row-3"
        >
          <CalendarDay
            dayNumber={15}
            key="calendarDay-15"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={16}
            key="calendarDay-16"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={17}
            key="calendarDay-17"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={18}
            key="calendarDay-18"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={19}
            key="calendarDay-19"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={20}
            key="calendarDay-20"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={21}
            key="calendarDay-21"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
        </tr>
        <tr
          key="calendar-row-4"
        >
          <CalendarDay
            dayNumber={22}
            key="calendarDay-22"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={23}
            key="calendarDay-23"
            mealScheduleForDay={
              Array [
                Object {
                  "assembly_id": null,
                  "created_at": "2019-09-04T01:00:35.645Z",
                  "id": 1,
                  "item_id": 15,
                  "meal_time": "none",
                  "order_item_id": null,
                  "purchased": false,
                  "schedule_date": "2019-09-24T00:00:00.000Z",
                  "updated_at": "2019-09-04T01:00:35.645Z",
                },
                Object {
                  "assembly_id": 2,
                  "created_at": "2019-09-04T01:00:35.645Z",
                  "id": 2,
                  "item_id": null,
                  "meal_time": "none",
                  "order_item_id": null,
                  "purchased": false,
                  "schedule_date": "2019-09-24T00:00:00.000Z",
                  "updated_at": "2019-09-04T01:00:35.645Z",
                },
              ]
            }
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={24}
            key="calendarDay-24"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={25}
            key="calendarDay-25"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={26}
            key="calendarDay-26"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={27}
            key="calendarDay-27"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={28}
            key="calendarDay-28"
            mealScheduleForDay={
              Array [
                Object {
                  "assembly_id": 2,
                  "created_at": "2019-09-03T01:00:35.645Z",
                  "id": 3,
                  "item_id": null,
                  "meal_time": "none",
                  "order_item_id": null,
                  "purchased": false,
                  "schedule_date": "2019-09-29T00:00:00.000Z",
                  "updated_at": "2019-09-07T01:00:35.645Z",
                },
              ]
            }
            onClickDayHandler={[Function]}
          />
        </tr>
        <tr
          key="calendar-row-5"
        >
          <CalendarDay
            dayNumber={29}
            key="calendarDay-29"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <CalendarDay
            dayNumber={30}
            key="calendarDay-30"
            mealScheduleForDay={Array []}
            onClickDayHandler={[Function]}
          />
          <td
            className="calendarTable-td"
            key="blankDay-0"
          >
            <div
              className="calendarTable-td-cell mix-calendarTable-td-cell_isEmpty"
            >
               
            </div>
          </td>
          <td
            className="calendarTable-td"
            key="blankDay-1"
          >
            <div
              className="calendarTable-td-cell mix-calendarTable-td-cell_isEmpty"
            >
               
            </div>
          </td>
          <td
            className="calendarTable-td"
            key="blankDay-2"
          >
            <div
              className="calendarTable-td-cell mix-calendarTable-td-cell_isEmpty"
            >
               
            </div>
          </td>
          <td
            className="calendarTable-td"
            key="blankDay-3"
          >
            <div
              className="calendarTable-td-cell mix-calendarTable-td-cell_isEmpty"
            >
               
            </div>
          </td>
          <td
            className="calendarTable-td"
            key="blankDay-4"
          >
            <div
              className="calendarTable-td-cell mix-calendarTable-td-cell_isEmpty"
            >
               
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
